### DISCUSSION TITLE
MPC VS Simulator results

### DISCUSSION THREAD

**Discussion by ThomasGiavasopoulos** [do-mpc/do-mpc#259] [Q&A]
Good afternoon,

I have observed that in the standard application of do-mpc, the **MPC** and **Simulator** results appear to be exactly the same.

Even when I set the step of the simulator to be larger, meaning that an input is applied for longer, the results still remain the same!?

I would expect the positions to be different for the simulator when the controller step is different. 

I save the results in this way:

`save_results([mpc, simulator])`

Is it safe to assume that the mpc and simulator results exactly the same? When would they be different?

Also!

When running the loop:

```
iteration_count = 5
    
    iter_list = np.linspace(1, iteration_count, iteration_count, dtype=int)
    
    for i in iter_list:
        
        u0 = mpc.make_step(x0)
        x0 = simulator.make_step(u0)
```

if I make a list of x0's manually and compare to the results from the MPC, they are different. Does the MPC class record slightly different x0's with time? 

I have observed that the difference is of the order to 1e-4 at the start and at the end and can be as high as 2 in the middle of the iter_list loop. I would expect that x0's collected manually and the MPC step results to be the same.

Thank you very much.

Thomas

**Comment by 4flixt** [comment]
Hi,

The state trajectory stored in the MPC controller is based on the ``x0`` that you pass with every ``mpc.make_step``. If you just obtain that from the simulator, without an additional estimator, the stored trajectories should be identical. 
This will, of course, also apply if you use different time-steps for simulator and controller.
What will be different in this case, is the predicted trajectory of the MPC controller and the obtained trajectory after applying the inputs for longer period of times will differ. 

Controller, estimator and Simulator have their own stored sequence of states (and inputs ...), because in the general case these sequences can be different: The simulator might use a different (high-fidelity) model of the process and the controller gets the estimated state from the estimator. 

Best,
4flixt

**Reply by ThomasGiavasopoulos** [reply]
Thank you very much for your reply. It is very helpful. 

To be clear, I have just discovered that when I manually created the list for x0, I had **not appended** the `first x0`, resulting in the difference I discussed above. It appears that now the `x0` match.

This extra investigation that I do results from a major problem that I am facing:

When running the following script:

```
iteration_count = 5
    
iter_list = np.linspace(1, iteration_count, iteration_count, dtype=int)

for i in iter_list:
  
  u0 = mpc.make_step(x0)
  x0[0] = x0[0]+5
  x0=x0
```

it appears that the MPC only stores 5 elements for each state. It also stores the initial state given. This means that the last state is not stored. The problem occurs when I demand that my solution loop stops when an x0 condition is surpassed (I assume that the last x0 at the end of the loop represents the final positions). The final state array is not stored in the `save_results([mpc])`. This means that I have to "invisibly" run the mpc.make_step one last time to record the final state. 

I assume that the problem is that the mpc.make_step only stores the `x0' s` and not the final state. Is this something that is true?

In the image below: 

<img width="664" alt="image" src="https://user-images.githubusercontent.com/57414585/163800683-f40501fc-736c-453d-92dd-a0bac94b8683.png">

50 is the initial state and 5 is added to it for 5 times. The final position should be 75 if the final input is executed, but it appears as 70, thus missing the final value.

Thank you again. Your help is invaluable @4flixt . 

Thomas
